docs/RFCS/data_structures_for_logical_planning.md

```diff
@@ -0,0 +1,1001 @@
+interpretation during query execution) remains, as before, the result
+of evaluating the top-level expression.
+
+A logical plan is *represented* (in code, in memory) by:
+
+- a single flat array of items that conceptually represent *classes* of expressions.
+- an index into the top-level array that identifies the root expression (class).
+- a class is itself a data structure comprised of:
+  - a member data structure that contains the *properties* of the class.
+  - a (dynamically allocated) unordered multiset of one or more
+    concrete expression representations, called "m-expressions".
+    You can think about an m-expression as a particular strategy
+    to compute the results.
+
+The data structure has a principal invariant: *all the m-expressions
+inside a class have the same semantics.* In particular, they share
+the same properties.
+
+The top-level array together with its item nodes is called a *memo*.
+
+The new thing in this RFC is that the representation of an expression
+is split (in code, in memory) into two places: one at the head of the class
+node where its properties are encoded, and one in the representation
+set where its structure (and, later, physical properties) is encoded.
+
+## Expression representations
+
+Before we look at examples we can enumerate the possible m-expression
+types we'll need for them:
+
+- `(literal "N")` to represent the literal value N.
+- `(+ x y)` where `x` and `y` are indexes into the memo.
+- `(= x y)` where `x` and `y` are indexes into the memo.
+- `(filter x y)` where `x` is an index into the memo that designates
+  where the filter is getting data from, and `y` is an index into the
+  memo that designates the predicate being computed for each row to
+  decide whether the row is omitted.
+- `(project x y)` where `x` is an index into the memo that designates
+  where the projection is getting data from, and `y` is an index into
+  the memo that designates the value being computed for each row.
+- `(var "name")`, which represents the value in column `name` in
+  the current row in the relational context where it is evaluated.
+- `(scan <name> x)` which extracts all the rows in the named table into
+  the var(s) indexed by `x`.
+- `(exists x)` which evaluates to non-zero iff the result of the evaluation
+  of the expression indexed by `x` has more than zero rows.
+
+### Examples
+
+Let's take an first simple example:
+
+```sql
+select x+42 from foo
+```
+
+Corresponding memo, with properties omitted:
+
+```
+root index: 0
+0: (project 1 3)
+1: (scan "foo" 2)
+2: (var "x")
+3: (+ 2 4)
+4: (literal "42")
+```
+
+We can conceptually visualize this memo as an expression tree by
+drawing the edges between index references and the representations;
+there are some non-tree additional edges between variable definitions
+and uses:
+
+```
+         (project)
+          /     \
+  (scan "foo")    (+)
+       |        /   \
+    (var "x")--'   (lit 42)
+```
+
```

Comment:
This double referencing of `var "x"` is still confusing to me (because I don't understand why it exists). I think my confusion can be addressed if you explain to me why I can't just remove the `(project)` root and its left outgoing edge, and make `(+)` the new root. Then you get a tree in which `scan foo` pushes data upwards, `var "x"` is basically the projection, then you get the addition and that is the result.

I think `project` acts like a `zip` in that it forms output rows from multiple inputs. 
But the name suggests the opposite, namely that it extracts individual components from something bigger (like a tuple). It has this characteristic because it's attached to the `scan`, but the scan also "pushes" into the `var`s, so I don't know why `(project)` even has to know about the data source and couldn't be a pure `zip`.

I assume there's a reason for that though, which I look forward to learning.

Reply:
Ok, I like the stacked projections example for that. Maybe the order could be switched up so that there's a nontrivial projection early in the process. Just a suggestion though.